catalog/core/catalog-core-api-impl/src/main/java/ddf/catalog/data/impl/MetacardImpl.java

```diff
@@ -156,10 +159,20 @@ public MetacardImpl(Metacard metacard, MetacardType type) {
             throw new IllegalArgumentException(
                     MetacardType.class.getName() + " instance should not be null.");
         }
-        map = new HashMap<>();
-        for (AttributeDescriptor attribute : metacard.getMetacardType().getAttributeDescriptors()) {
-            map.put(attribute.getName(), metacard.getAttribute(attribute.getName()));
+
+        if (metacard.getSourceId() != null) {
+            setSourceId(metacard.getSourceId());
         }
+
+        map = metacard.getMetacardType()
+                .getAttributeDescriptors()
+                .stream()
+                .filter(Objects::nonNull)
+                .filter(attributeDescriptor -> attributeDescriptor.getName() != null)
+                .filter(attributeDescriptor -> metacard.getAttribute(attributeDescriptor.getName())
+                        != null)
+                .map(AttributeDescriptor::getName)
```

Comment:
This is cleaner, but could be even a bit better by moving this `map` up two lines:

```java
//...
.filter(Objects::nonNull)
.map(AttributeDescriptor::getName)
.filter(Objects::nonNull)
.filter(attrName -> metacard.getAttribute(attrName) != null)
.collect(//...
```